docs/controller-guidelines.md

```diff
@@ -539,189 +543,178 @@ A controller should define and export a type union that holds all of its events.
 >;
 ```
 
-If, in a test, you need to access all of the actions included in a controller's messenger allowlist, use the [`ExtractAvailableAction` utility type](../packages/base-controller/tests/helpers.ts):
+If, in a test, you need to access all of the actions supported by a messenger, use the [`MessengerActions` utility type](../packages/messenger/src/Messenger.ts):
 
 ```typescript
-// NOTE: You may need to adjust the path depending on where you are
-import { ExtractAvailableAction } from '../../base-controller/tests/helpers';
+import type { MessengerActions, MessengerEvents } from '@metamask/messenger';
 
 const messenger = new Messenger<
-  ExtractAvailableAction<FooControllerMessenger>,
-  never
+  controllerName,
+  MessengerActions<FooControllerMessenger>,
+  MessengerEvents<FooControllerMessenger>
 >();
 ```
 
 ## Define, but do not export, a type union for external event types
 
-A controller may wish to subscribe to events defined by other controllers, and therefore will need to define them in the messenger's allowlist.
+A controller may wish to subscribe to events defined by other controllers, and therefore will need to include them in the controller messenger's type definition.
 
-In this case, the controller should group these types into a type union so that they can be easily passed to the `RestrictedMessenger` type. However, it should not export this type, as it would then be re-exporting types that another package has already exported.
+In this case, the controller should group these types into a type union so that they can be easily passed to the `Messenger` type. However, it should not export this type, as it would then be re-exporting types that another package has already exported.
 
 The name of this type should be `AllowedEvents`.
 
-This type should not only be passed to `RestrictedMessenger` as the 3rd type parameter, but should also be included in its 5th type parameter.
+This type should be passed to `Messenger` as part of the 3rd type parameter, in a type union with internal events.
 
-🚫 **`never` is passed as the 5th type parameter**
+🚫 **`AllowedEvents` is exported**
 
 ```typescript
+/* === packages/foo-controller/src/FooController.ts === */
+export type FooControllerEvents =
+  | FooControllerMessageReceivedEvent
+  | FooControllerNotificationAddedEvent;
+
 export type AllowedEvents =
   | BarControllerSomethingHappenedEvent
   | BarControllerSomethingElseHappenedEvent;
 
-export type FooControllerMessenger = RestrictedMessenger<
+export type FooControllerMessenger = Messenger<
   'FooController',
-  never,
-  AllowedEvents,
-  never,
-  never
+  FooControllerActions,
+  FooControllerEvents | AllowedEvents
 >;
+
+/* === packages/foo-controller/src/index.ts === */
+
+export type { AllowedEvents } from '@metamask/foo-controller';
 ```
 
-🚫 **`AllowedEvents['type']` is passed as the 5th type parameter, but `AllowedEvents` is exported**
+🚫 **External events are included in controller actieventn type**
```

Comment:
Typo:

``` suggestion
🚫 **External events are included in controller event type**

Reply:
But also, perhaps this anti-pattern should be moved up to "Define and export a type union for internal event types"?

Reply:
Oh? I think it makes more sense in this section about external events, since it's an example of putting these external events in the wrong place.

Maybe it would be simpler to combine these two sections? But it does seem like this example belongs here unless we do combine them.

Reply:
Hmm, interesting. Perhaps I'll do another pass of this whole document later. Still, this is a good addition, so I approve.